### PR TITLE
#167727731 Get and update user details

### PR DESCRIPTION
**Pull request by Mcdavid95**
#### What does this PR do?
- Adds endpoint to get user
- Adds endpoint to update user

#### Description
- update user model with new attributes
- update user sequelize file with new attributes
- add get and update service
- add endpoint to get and update user details
- add middleware to check jwt token
- add middleware to check user id

#### How can this be manually tested
- Clone the repo
- Install dependencies, setup database and run `npm start:dev`
- Using postman or any http request tool, make a get request to `api/v1/users/:userId` to get user details
- Using postman or any http request tool, make a put request to `api/v1/users/:userId` and pass in the values to update user details

#### Relevant Pivotal tracker stories
- [#167727731](https://www.pivotaltracker.com/story/show/167727731)

### CODE REVIEW

File: src/controllers/userController.js [hunk 1]
```diff
@@ -43,6 +44,63 @@ const signUp = async (req, res) => {
...
+ * @description get's details of logged in user
+ */
+const getUserDetailsById = async (req, res) => {
+  try {
+    const { userId } = req.params;
+    const options = { attributes: { exclude: ['password'] } };
+    const user = await getById(User, userId, options);
+    return response(res, 200, 'success', { user });
+  } catch (error) {
+    return response(res, 500, 'error', { message: 'Internal server error' });
```
Comment (Mcdavid95) [inline src/controllers/userController.js]: message should be in message object

File: src/controllers/userController.js [hunk 2]
```diff
@@ -43,6 +44,63 @@ const signUp = async (req, res) => {
...
+ * @param {Object} res - server response
+ * @returns {Object} - custom response
+ * @description updates details of logged in user
+ */
+const updateUserDetails = async (req, res) => {
+  try {
+    const { userId } = req.params;
+    const { id } = req.decoded;
+    if (userId !== id) {
+      return response(res, 403, 'error', { message: 'You are not authorized to edit this profile' });
```
Comment (Mcdavid95) [inline src/controllers/userController.js]: add message to message object

File: src/controllers/userController.js [hunk 3]
```diff
@@ -43,6 +44,63 @@ const signUp = async (req, res) => {
...
+      preferredCurrency,
+      gender,
+      role,
+      lineManager,
+      currentLocation
+    };
+    const updatedUser = await update(User, userId, userInfo);
+    return response(res, 200, 'success', { updatedUser });
+  } catch (error) {
+    return response(res, 500, 'error', { message: 'Internal server error' });
```
Comment (Mcdavid95) [inline src/controllers/userController.js]: add message to message object

File: src/controllers/userController.js [hunk 4]
```diff
@@ -43,6 +44,63 @@ const signUp = async (req, res) => {
...
+      phoneNo,
+      birthDate,
+      preferredLanguage,
+      preferredCurrency,
+      gender,
+      role,
+      lineManager,
+      currentLocation
+    };
+    const updatedUser = await update(User, userId, userInfo);
```
Comment (ladaposamuel) [inline src/controllers/userController.js]: This update function is not descriptive enough

File: src/middlewares/userMiddlewares.js [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+   * @param {function} next next function
+   * @returns {object} custom response
+   * @description checks if userId passed to params is valid
+   */
+export const checkUserId = async (req, res, next) => {
+  try {
+    const { userId } = req.params;
+    const user = await getById(User, userId);
+    if (!user) {
+      return response(res, 404, 'error', 'User not found');
```
Comment (Mcdavid95) [inline src/middlewares/userMiddlewares.js]: use message object

File: src/middlewares/userMiddlewares.js [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+ * @returns {Object} response object
+ */
+export const verifyToken = async (req, res, next) => {
+  const token = req.headers['x-access-token'];
+  try {
+    if (token) {
+      const decoded = await jwt.verify(token, process.env.SECRET_KEY);
+      const user = await getById(User, decoded.id);
+      if (!user) {
+        return response(res, 403, 'error', { message: 'Token you provided is invalid' });
```
Comment (Mcdavid95) [inline src/middlewares/userMiddlewares.js]: change status code to unauthorized(401)

File: src/middlewares/userMiddlewares.js [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+};
+/**
+ * @method hasToken
+ * @param {*} req
+ * @param {*} res
+ * @param {*} next
+ * @returns {Object} response object
+ */
+export const verifyToken = async (req, res, next) => {
+  const token = req.headers['x-access-token'];
```
Comment (Mcdavid95) [inline src/middlewares/userMiddlewares.js]: Change to `Authorization`

File: src/middlewares/userMiddlewares.js [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    if (token) {
+      const decoded = await jwt.verify(token, process.env.SECRET_KEY);
+      const user = await getById(User, decoded.id);
+      if (!user) {
+        return response(res, 403, 'error', { message: 'Token you provided is invalid' });
+      }
+      req.decoded = decoded;
+      return next();
+    }
+    return response(res, 403, 'error', { message: 'You have to be logged in' });
```
Comment (Mcdavid95) [inline src/middlewares/userMiddlewares.js]: use message object

File: src/routes/api/user.js [hunk 9]
```diff
@@ -49,5 +52,69 @@ const userRoute = (router) => {
...
+   *       - Users
+   *     summary: Get one user's details
+   *     parameters:
+   *       - in: params
+   *         name: userId
+   *         schema:
+   *           type: string
+   *         required: true
+   *       - in: header
+   *         name: x-access-token
```
Comment (Mcdavid95) [inline src/routes/api/user.js]: change to Authorizaton

File: src/services/dbServices.js [hunk 10]
```diff
@@ -1,9 +1,36 @@
...
+  /**
    * Database create service funcion
    * @param {Object} model - Defined model
    * @param {Object} data - Data to be created
    * @returns {Promise} - Promise response
    */
-const create = (model, data) => model.create(data);
+  create(model, data) {
+    return model.create(data);
+  },
```
Comment (Mcdavid95) [inline src/services/dbServices.js]: add empty line

**Comment by tejiri4** [comment]
Write test for your file changes and fix the code climate issues

File: src/controllers/userController.js (hunk 4 above)
Comment (Mcdavid95) [reply src/controllers/userController.js]: can you please shed more light on this.

**Comment by Mcdavid95** [comment]
> Write test for your file changes and fix the code climate issues

done

File: src/middlewares/userMiddlewares.js [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+};
+/**
+ * @method verifyToken
+ * @param {*} req
+ * @param {*} res
+ * @param {*} next
+ * @returns {Object} response object
+ */
+export const verifyToken = async (req, res, next) => {
+  const token = req.headers.authorization;
```
Comment (tejiri4) [inline src/middlewares/userMiddlewares.js]: Use a Split function to separate Bearer from the token

**Comment by tejiri4** [comment]
When I update a user the password is sent back, kindly remove the password from the response body. Phone number and birth date is not properly validated when updating a user it  and rebase  with develop when you done

File: src/middlewares/userMiddlewares.js [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+export const checkToken = async (req, res, next) => {
+  let token = req.headers.authorization;
+  if (token && token.startsWith('Bearer ')) {
+    // Remove Bearer from string
+    const [, jwtToken] = token.split(' ');
+    token = jwtToken;
+  }
+  try {
+    if (token) {
+      const decoded = await verifyToken(token);
```
Comment (tejiri4) [inline src/middlewares/userMiddlewares.js:52]: welldone when rebasing do maintain this your function and replace with develop

**Review by tejiri4 — APPROVED** [review]
LGTM

**Pull request merged**